### PR TITLE
Ensure response body is collected on the EventLoop

### PR DESCRIPTION
**Pull request by 0xTim**
This ensures that the request body is collected on the correct `EventLoop` no matter where if was called from. This avoids hitting an assertion failure when calling 
this from a Swift Concurrency context.

### CODE REVIEW

**Comment by 0xTim** [comment]
cc @weissi

File: Sources/Vapor/HTTP/BodyStream.swift [hunk 1]
```diff
@@ -45,7 +45,13 @@ public protocol BodyStreamWriter {
 extension BodyStreamWriter {
     public func write(_ result: BodyStreamResult) -> EventLoopFuture<Void> {
         let promise = self.eventLoop.makePromise(of: Void.self)
-        self.write(result, promise: promise)
+        if self.eventLoop.inEventLoop {
```
Comment (weissi) [inline Sources/Vapor/HTTP/BodyStream.swift:48]: @0xTim this isn't 100% correct: `inEventLoop` is "best effort": if it says we're on the EventLoop that's true but the contrary isn't necessarily true. Some EventLoop implementations might not be able to reliably determine if you're on them.
So this'll work just fine for Vapor today which always uses `MultiThreadedEventLoopGroup` which can reliably determine this but I'd rewrite it similar to this

```
    func write(_ result: BodyStreamResult, promise: EventLoopPromise<Void>?) {
        func write0() {
            switch result {
            case .buffer(var buffer):
                self.buffer.writeBuffer(&buffer)
            case .error(let error):
                self.promise.fail(error)
            case .end:
                self.promise.succeed(self.buffer)
            }
            promise?.succeed(())
        }

        if self.eventLoop.inEventLoop {
            write0()
        } else {
            self.eventLoop.execute { write0() }
        }
    }
```

which works regardless and just hits the "slow path" if the EventLoop can't reliably determine that we're on the EL already.

**Review by gwynne — DISMISSED** [review]
What @weissi said 😅

File: Sources/Vapor/HTTP/BodyStream.swift (hunk 1 above)
Comment (0xTim) [reply Sources/Vapor/HTTP/BodyStream.swift]: @weissi just to be clear is the difference that the promise is fulfilled on the event loop in your example but not in mine? Updated!

Comment (weissi) [reply Sources/Vapor/HTTP/BodyStream.swift]: doens't matter where the promise is fulfilled. The issue is that this some other EventLoop implementations (none that I would be aware of in the real world right now) might always return `false` from `eventLoop.inEventLoop`. And I think your original code would then up in an infinite loop if that were to happen.

**Review by weissi — APPROVED** [review]
Thank you, LGTM

**Comment by 0xTim** [comment]
Added to #3055

**Pull request closed without merging**